systray api on osx

fffw opened the pull request:
getlantern/lantern#2143 for discussion on code.

fffw commented:
Haven't found way to wait for `applicationDidFinishLaunching:` before any systray operation.
The caller in goroutine and runloop may runs in the same thread, so we can't just wait for a NSCondition.

oxtoacart commented:
> Haven't found way to wait for applicationDidFinishLaunching: before any systray operation.

Will this work? https://github.com/getlantern/systray/tree/issue-2143-ox

It seems to work in my testing...

fffw commented:
@ox You brilliant! the `onReady()` callback is simple and beautiful.




I’ll fix a few memory leaks in obj-c code.

oxtoacart commented:
@fffw You're too kind!

oxtoacart commented on systray.go:
```diff
@@ -0,0 +1,139 @@
+// overwrites the first.
+//
+// AddMenuItem can be safely invoked from different goroutines.
+func AddMenuItem(id string, title string, tooltip string) *MenuItem {
```
@fffw I don't like that the library require me to manage ids for the menu items, and I don't understand why it's necessary.  You could have AddMenuItem assign the id internally as a private attribute.

fffw replied:
Yeah great insight

oxtoacart commented on systray.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// Update propogates changes on a menu item to systray
+func Update(item *MenuItem) {
```
I think this would be more convenient if it were

`func (item *MenuItem) Update()`

fffw replied:
I want the name to emphasis the idea that we are changing the actual menu item according to the value of `item`, rather than update itself. `Update` might not be a good name. `item.Redraw()`?

oxtoacart replied:
Oh, I see what you're saying.  Changing properties on a MenuItem doesn't actually affect the UI until this update method is called.  Maybe the API would be cleaner if the properties on the MenuItem weren't public and it instead exposed mutator methods that automatically send the update to the UI, for example:

```go
func (item *MenuItem) SetTitle(title string) {
}

func (item *MenuItem) SetTooltip(tooltip string) {
}
```
func (item *MenuItem) SetEnabled(enabled bool) {
}

func (item *MenuItem) SetChecked(checked bool) {
}

oxtoacart commented on example/iconwin.go:
```diff
@@ -0,0 +1,367 @@
+
+package main
+
+var iconData []byte = []byte {
```
Can you please move iconwin.go and iconunix.go into their own folders?  I got confused because there are multiple main files in this same folder, and when I run `go run *.go` I get an error.

fffw replied:
Nice idea!

oxtoacart commented on example/main.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/skratchdot/open-golang/open"
+)
+
+func main() {
```
Very nice example!

oxtoacart commented on example/main.go:
```diff
@@ -0,0 +1,62 @@
+		mEnabled := systray.AddMenuItem("enable", "Enabled", "Enabled")
+		systray.AddMenuItem("ignore", "Ignored", "Ignored")
+		mUrl := systray.AddMenuItem("lantern", "Open Lantern.org", "my home")
+		mQuit := systray.AddMenuItem("quit2", "退出", "Quit the whole app")
```
I like the inclusion of non-Latin characters!

oxtoacart commented on systray.go:
```diff
@@ -0,0 +1,139 @@
+// SetIcon sets the systray icon.
+// iconBytes should be the content of .ico for windows and .png for other
+// platforms.
+func SetIcon(iconBytes []byte) {
```
It would be convenient if this worked with one type of file for all platforms (preferably png).  Something like [icoutils](http://www.nongnu.org/icoutils/) could be used to convert PNG to ICO on the fly.

oxtoacart replied:
Actually, never mind.  It's not worth the work of integrating another library just for this.

fffw replied:
Mac and Linux does support ico as trayicon. I'll update doc.

oxtoacart replied:
Oh, that's perfect, thanks!

fffw commented:
@oxtoacart improved per your advice.

A few concerns come into mind:

* Windows and AppIndicator API only allows set icon directly from file name, so I extract icon bytes to a temporary file, load it as icon, and remove it. It's a waste of resource, but acceptable if `setIcon()` is not invoked frequently.

* As we can't explicitly ask a goroutine to run in seperate thread, it's hard to verify if the platform dependent code is really thread-safe.

oxtoacart commented:
> Windows and AppIndicator API only allows set icon directly from file name, so I extract icon bytes to a temporary file, load it as icon, and remove it. It's a waste of resource, but acceptable if setIcon() is not invoked frequently.

If we wanted to be slightly more efficient, we could have SetIcon accept a path to a file and let the program manage the temp file.  That way if we're doing something like toggling between a couple of icons (which Lantern will do to indicate proxying status), we could just keep both images on disk without having to constantly recreate the temp files.

> As we can't explicitly ask a goroutine to run in seperate thread, it's hard to verify if the platform dependent code is really thread-safe.

Yeah. Thread safety is really difficult to test anyway.  If it were pure go, you could look for race conditions by running with the `-race` flag, but I don't know if that will be useful when using CGO.

oxtoacart commented on systray.go:
```diff
@@ -0,0 +1,178 @@
+// Don't create it directly, use the one systray.AddMenuItem() returned
+type MenuItem struct {
+	// Ch is the channel which will be notified when the menu item is clicked
+	Ch chan interface{}
```
Small niggle - might want to call this `ClickedCh` or something like that.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,248 @@
+#include "systray.h"
+
+#define WM_SYSTRAY_MESSAGE (WM_USER + 1)
+#define MAX_LOADSTRING 100
```
This constant doesn't seem to be used anywhere.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,248 @@
+#include <shellapi.h>
+#include "systray.h"
+
+#define WM_SYSTRAY_MESSAGE (WM_USER + 1)
```
Why WM_USER+1?

fffw replied:
It's a custom? WM_USER is also ok.

oxtoacart replied:
Sorry, I just don't understand why our constant is related to WM_USER, or what WM_USER is.  It's probably just because I'm not a windows programmer, but a comment explaining the relationship might be helpful.

fffw replied:
excerpt from [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/ms644931(v=vs.85).aspx)

>  Message numbers in the second range (WM_USER through 0x7FFF) can be defined and used by an application to send messages within a private window class.

It's the way the systray area notifying our application's main message loop. I'll add a comment.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,248 @@
+{
+	int textlen;
+	wchar_t * result;
+	textlen = MultiByteToWideChar(CP_UTF8, 0, str, -1, NULL ,0);
```
Need to handle error (i.e. textlen == 0). See https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072%28v=vs.85%29.aspx.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,248 @@
+	int textlen;
+	wchar_t * result;
+	textlen = MultiByteToWideChar(CP_UTF8, 0, str, -1, NULL ,0);
+	result = (wchar_t *)calloc((textlen+1), sizeof(wchar_t));
```
Need to handle error (i.e. result == 0).

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,248 @@
+	wchar_t * result;
+	textlen = MultiByteToWideChar(CP_UTF8, 0, str, -1, NULL ,0);
+	result = (wchar_t *)calloc((textlen+1), sizeof(wchar_t));
+	MultiByteToWideChar(CP_UTF8, 0, str, -1, (LPWSTR)result, textlen);
```
Need to check return value and handle error.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,248 @@
+void ShowMenu(HWND hWnd)
+{
+	POINT p;
+	GetCursorPos(&p);
```
Need to check return value and handle error.  See https://msdn.microsoft.com/en-us/library/windows/desktop/ms648390%28v=vs.85%29.aspx.

I guess in general, you need to come up with a scheme for checking return values and handling errors.

oxtoacart commented on systray_darwin.m:
```diff
@@ -0,0 +1,175 @@
+  [menuId release];
+  [title release];
+  [tooltip release];
+  [super dealloc];
```
@fffw I think it might make sense to switch to using [ARC](https://developer.apple.com/library/mac/releasenotes/ObjectiveC/RN-TransitioningToARC/Introduction/Introduction.html#//apple_ref/doc/uid/TP40011226) for all of our Objective C code.  I think that'll make it easier for everyone to maintain.  What do you think?

fffw replied:
I'm not sure if I know enough about ARC, but I'll try, thanks.

oxtoacart replied:
I think the main thing is to switch the instance variables of MenuItem and AppDelegate to be Properties, get rid of the dealloc method and add the -fobjc-arc compiler flag.  Also get rid of the (assign) qualifier on line 48 and the release statements in your mutators for setIcon, etc.  I think you might also be able to get rid of the autorelease on line 64, but I'm not sure about that.

oxtoacart replied:
BTW, I'm not 100% sure that this is worth doing.  I just brought it up because it seems to be Apple's suggested way of doing things these days.

oxtoacart commented on systray_linux.c:
```diff
@@ -0,0 +1,139 @@
+	char* temp_file_name = malloc(PATH_MAX);
+	g_array_append_val(global_temp_icon_file_names, temp_file_name);
+	strcpy(temp_file_name, "/tmp/systray_XXXXXX");
+	int fd = mkstemp(temp_file_name);
```
Need to check for error -1.  http://man7.org/linux/man-pages/man3/mkstemp.3.html

oxtoacart commented on systray_linux.c:
```diff
@@ -0,0 +1,139 @@
+
+gboolean do_set_icon(gpointer data) {
+	GBytes* bytes = (GBytes*)data;
+	if (global_temp_icon_file_names == NULL) {
```
Why allocate this lazily instead of in nativeLoop()?

fffw replied:
Yes, should in nativeLoop()

oxtoacart commented:
@fffw The Cocoa code looks good to me, other than that you may want to switch to using automatic reference counting.

The Windows and the Linux code need a lot more error handling.  If you want to go ahead and add that, I'll take a final look at that code afterwards.

oxtoacart commented:
So, I finally installed the equinox tool.  It's actually quite annoying.  In particular, it seems to insist on doing the compilation for me and then immediately uploads the binary, without giving me a chance to do the native code signing.  So, I'm not sure how we'll make it work.

fffw commented:
Thanks @oxtoacart for detailed review!
Followed your suggestion, I switched Cocoa code to ARC, and `leaks` report is good.
I also added a few error checking, just to prevent program from crash or undefined behaviour, error reporting is not elegant. As almost every Windows system call can fail, I ignored those don't harm.

oxtoacart commented on README.md:
```diff
@@ -1,15 +1,48 @@
-- https://github.com/merlinran/systray/compare/tray-menu-for-osx#diff-b7913550f26df0fe93932596ef6086c2R56
-- http://th30z.blogspot.com/2008/10/cocoa-system-statusbar-item-aka-traybar_2086.html (thanks @atavism)
-- http://golang.org/cmd/cgo/
-- https://code.google.com/p/go-wiki/wiki/LockOSThread
```
Why did you take out these references? I think it's good to keep them in.  If any of the Linux or Windows code was based on other examples, please include references to these as well.

fffw replied:
My code is mainly stolen from the two projects under Credits section. Though lots of documents and discussion threads were consulted, they may have little help to users of this library, so I shorten them.

if we list them all here.

oxtoacart replied:
Oh, I see - sorry, I missed the credits section since it wasn't changed in this PR.

oxtoacart commented on systray_linux.c:
```diff
@@ -0,0 +1,148 @@
+// runs in main thread, should always return FALSE to prevent gtk to execute it again
+gboolean do_set_icon(gpointer data) {
+	GBytes* bytes = (GBytes*)data;
+	char* temp_file_name = malloc(PATH_MAX);
```
Because this is allocated on the heap, it will leak.  You could just use `char temp_file_name[PATH_MAX]`.

oxtoacart replied:
Oh, sorry, I forgot that you're adding it to global_temp_icon_file_names.  What is the point of global_temp_icon_file_names?

fffw replied:
to remove created temp files when quit the app.

oxtoacart replied:
Ahh, so you have to keep the temp files around as long as the icon is in use?

oxtoacart replied:
Or can you delete the temp file once the icon has been loaded?

fffw replied:
Yes, we have to keep it.

fffw replied:
Let the caller to provide file name would make our code simpler, but that only transfers the work to caller, so I leave it here at this point.

oxtoacart replied:
Cool, thanks.  A comment in the code to explain the use of global_temp_icon_file_names might be useful.

fffw replied:
Comment added, thanks.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+		reportWindowsError("get tray menu position");
+		return;
+	};
+	SetForegroundWindow(hWnd); // Win32 bug work-around
```
What if this returns false?

fffw replied:
> If the window was brought to the foreground, the return value is nonzero.
> If the window was not brought to the foreground, the return value is zero.

Maybe the menu not shown in rarely circumstances? Actually, I just copied it from [trayhost project](https://github.com/cratonica/trayhost/blob/350dc33e9265ef5979fe6a0bd6091337197c7138/platform/windows/tray.c#L166).

oxtoacart replied:
Yeah, I guess that's alright.  That trayhost code has almost no error handling?!  That's a good reminder to me to closely inspect any C libraries that we want to use :)

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam) {
+	switch (message) {
+		case WM_MENUCOMMAND:
+			systray_menu_item_selected(GetMenuItemId(wParam));
```
If GetMenuItemId fails, it will return null.  If we pass null to systray_menu_item_selected, our call to `C.GoString` will either fail or return an empty string, in which case our menuItems[id] lookup fill return nil and our attempt to submit to menuItem.ClickedCh will fail.

I think we need to check for a NULL return value here and if so, don't call systray_menu_item_selected.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+	}
+
+	ShowWindow(hWnd, nCmdShow);
+	UpdateWindow(hWnd);
```
What if this returns false?

fffw replied:
Interestingly, I've never seen code checking this. The calling sequence and even the function names are not changed since Windows 3.x.

I think we don't need to take much care on the bootstrapping code, one call fails and the app will not able to startup.

oxtoacart replied:
Okay.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+
+
+BOOL createMenu() {
+	hTrayMenu = CreatePopupMenu();
```
Need to check for NULL here.

fffw replied:
SetMenuInfo will fail anyway ;-)

oxtoacart replied:
Will it fail or will it crash because of a NULL pointer?

fffw replied:
Windows will always check any parameter passed in system calls, or will crash the OS instead.

oxtoacart replied:
Gotcha.  As you can probably tell, I've never done any Windows programming :)  So, I'm just thinking like an inexperienced C programmer.

fffw replied:
You are so lucky not to touch Windows programming at all...

oxtoacart replied:
So I'm learning!  Really the only C/C++ GUI programming I ever did was BeOS, which was a joy.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+	wcex.lpszClassName  = szWindowClass;
+	wcex.hIconSm        = LoadIcon(NULL, IDI_APPLICATION);
+
+	RegisterClassEx(&wcex);
```
Need to check for 0 return value.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+	wcex.hbrBackground  = (HBRUSH)(COLOR_WINDOW+1);
+	wcex.lpszMenuName   = 0;
+	wcex.lpszClassName  = szWindowClass;
+	wcex.hIconSm        = LoadIcon(NULL, IDI_APPLICATION);
```
If any of these LoadIcon or LoadCursor calls fails, do we need to react to that somehow?

fffw replied:
I believe they rarely fail or not fail at all, as we pass them system predefined values. If they do fail, subsequent calls will also fail and window handle will not be created.

oxtoacart replied:
Okay.

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+
+		FILE* fIcon = _wfopen(szTempFileName, TEXT("wb"));
+		if (length != fwrite(iconBytes, 1, length, fIcon)) {
+			printf("error write temp icon file\n");
```
If we couldn't write the file, we should probably just return here, right?

oxtoacart commented on systray_windows.c:
```diff
@@ -0,0 +1,255 @@
+		};
+		fclose(fIcon);
+
+		hIcon = LoadImage(NULL, szTempFileName, IMAGE_ICON, 64, 64, LR_LOADFROMFILE);
```
Need to check for null here.

fffw commented:
@oxtoacart is it ok to merge this one and move on?

oxtoacart commented:
I've finished reviewing the code and tested on OS X.  I still need to build and test on Windows and Linux.  Will do that today.

oxtoacart commented:
Okay, Windows build and test on XP was successful.  Now just need to test Linux.

oxtoacart commented:
Okay, Ubuntu looks good too, merging!

oxtoacart commented:
Nice work!
